### PR TITLE
Consistency in Modulo operator

### LINKED ISSUE #508: Int modulo
`Int modulo(Int)`, defined in `math.ooc`, behaves oddly.

```ooc
import math

(5 % 3) toString() println() // outputs 2.
(5 % -3) toString() println() // also outputs 2.
(-5 % 3) toString() println() // outputs -2, but should output 1. 
// Too bad, but that's how the operator works in C.
((-5) % 3) toString() println() // also outputs -2.
(-5 % -3) toString() println() // also outputs -2.

(5 modulo(3)) toString() println() // outputs 2, as expected.
(5 modulo(-3)) toString() println() // outputs 1. Is this correct?
(-5 modulo(3)) toString() println() // outputs -2, but it should output 1. 
// Unlike the % operator, modulo() was written explicitly to handle this situation,
// but it still doesn't work correctly.
((-5) modulo(3)) toString() println() // outputs 1, which is what we want.
// So... it's not a bug in the `modulo` function itself.
(-5 modulo(-3)) toString() println() // outputs -1. Is this correct?
((-5) modulo(-3)) toString() println() // outputs 2. Is this correct?
```

We're not going to change the behavior of the `%` operator, but the `modulo` function that we added should work the way we want it to. Currently, it doesn't for `-5 modulo(3)`, but what about the cases where the divisor is negative? Should `modulo` always return a positive number?

### LINKED ISSUE #768: Inconsistent modulo operator
The `modulo` function in `Int`, `Int64`, and `Float` (and possibly others) are all inconsistent with each other. They should be corrected to comply with the following C# code from the original kean:
```csharp
public static int Modulo(int dividend, int divisor)
{
	if (dividend < 0)
		dividend += Integer.Ceiling(Integer.Absolute(dividend) / (float)divisor) * divisor;
	return dividend % divisor;
}
```
See the original discussion here: #508 

The ooc methods should *not* be static, though.

### PR DESCRIPTION
**Pull request by marcusnaslund**
Fixes #768

### CODE REVIEW

**Comment by marcusnaslund** [comment]
@simonmika or @davidhesselbom , will one of you peer review this?

**Comment by simonmika** [comment]
Could we have tests on this?

**Comment by simonmika** [comment]
I can peer review this.

**Comment by marcusnaslund** [comment]
Test added.

File: test/sdk/MathTest.ooc [hunk 1]
```diff
@@ -0,0 +1,30 @@
+use ooc-unit
+import math
+
+MathTest: class extends Fixture {
+	init: func {		
+		super("Math")
+		tolerance := 1.0e-5f		
+		this add("Modulo", func {
+			intValuePos := 22
```
Comment (simonmika) [inline test/sdk/MathTest.ooc]: I would love to also have tests where the dividend is closer to zero then the divider.

**Comment by simonmika** [comment]
I'm done with the peer review.

File: test/sdk/MathTest.ooc (hunk 1 above)
Comment (marcusnaslund) [inline test/sdk/MathTest.ooc]: Alright, I will add some more.

**Comment by marcusnaslund** [comment]
Ran into some rebasing problems earlier. Everything fixed now, but please peer review again @simonmika

File: source/plot/Axis.ooc [hunk 2]
```diff
@@ -116,9 +116,9 @@ Axis: class {
 	getFirstTickValue: func -> Float {
 		result: Float
 		if ((this min < this tick && this min > 0.0f) || (this min > this tick && this min < 1.0f))
-			result = this min + (this tick - Float modulo(this min, this tick))
+			result = this min + (this tick - (this min modulo(this tick)))
```
Comment (simonmika) [inline source/plot/Axis.ooc]: Method calls have higher precedence then operators, should be:
```ooc
 result = this min + (this tick - this min modulo(this tick))
```

Comment (simonmika) [inline source/plot/Axis.ooc]: By the way, somebody should fix the names `minimum` and `maximum`, but that is hardly the responsibility if this PR.

File: source/plot/Axis.ooc [hunk 3]
```diff
@@ -116,9 +116,9 @@ Axis: class {
...
 		else
-			result = this min - Float modulo(this min, this tick)
+			result = this min - (this min modulo(this tick))
```
Comment (simonmika) [inline source/plot/Axis.ooc]: Same as above.

File: source/plot/SvgWriter2D.ooc [hunk 4]
```diff
@@ -79,7 +79,7 @@ SvgWriter2D: class {
 					numPlotsX = 2
 				else
 					numPlotsX = 3
-				numPlotsY = Int modulo(this svgPlots count, numPlotsX) ? 1 + this svgPlots count / numPlotsX : this svgPlots count / numPlotsX
+				numPlotsY = this svgPlots count modulo(numPlotsX) ? 1 + this svgPlots count / numPlotsX : this svgPlots count / numPlotsX
```
Comment (simonmika) [inline source/plot/SvgWriter2D.ooc]: You didn't introduce it but we should probably not do it the C way and assume that integers have a boolean value. It is probably enough to do the lazy version for not null.

Comment (simonmika) [inline source/plot/SvgWriter2D.ooc]: `numberOfPlotsY` or something. Again hardly something for this PR.

File: source/plot/SvgWriter2D.ooc [hunk 5]
```diff
@@ -87,7 +87,7 @@ SvgWriter2D: class {
 			plotSize := FloatVector2D new(this size x / numPlotsX, this size y / numPlotsY)
 			position := FloatPoint2D new()
 			for (i in 0 .. this svgPlots count) {
-				position x = plotSize x * Int modulo(i, numPlotsX)
+				position x = plotSize x * (i modulo(numPlotsX))
```
Comment (simonmika) [inline source/plot/SvgWriter2D.ooc]: To many parentheses.

File: source/plot/SvgWriter2D.ooc (hunk 4 above)
Comment (marcusnaslund) [inline source/plot/SvgWriter2D.ooc]: You mean `this svgPlots count modulo(numPlotsX) != 0`? I agree.

File: source/plot/SvgWriter2D.ooc [hunk 6]
```diff
@@ -87,7 +87,7 @@ SvgWriter2D: class {
...
 				position y = plotSize y * (i / numPlotsX)
```
Comment (simonmika) [inline source/plot/SvgWriter2D.ooc]: Here there are to many parenthesis again although not exactly on you (again).

File: source/sdk/math.ooc [hunk 7]
```diff
@@ -60,7 +60,7 @@ extend Short {
 }
 
 extend Int64 {
-	modulo: func (divisor: This) -> This { 
+	modulo: func (divisor: This) -> This {
 		result := this - (this / divisor) * divisor
```
Comment (simonmika) [inline source/sdk/math.ooc]: To many parenthesis again.

File: source/sdk/math.ooc [hunk 8]
```diff
@@ -68,14 +68,10 @@ extend Int64 {
 
 extend Int {
 	modulo: func (divisor: This) -> This {
-		if (divisor < 0) {
-			this *= -1
-			divisor *= -1
-		}
-		result := this % divisor
+		result := this
```
Comment (simonmika) [inline source/sdk/math.ooc]: It is not strictly necessary to introduce `result` as you could as well modify `this` to contain your result as this is a call by value member of a cover but this way is probably much clearer.

File: source/sdk/math.ooc [hunk 9]
```diff
@@ -306,7 +299,7 @@ extend Float {
 		value
 	}
 	minusPiToPiOverTwo: static func (value: This) -> This {
```
Comment (simonmika) [inline source/sdk/math.ooc]: Again, not on you, but if we really should have this method it should probably be called `minusPiToPiHalf` and not be static, which is true for the methods above as well.

File: test/sdk/MathTest.ooc [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+	init: func {
+		super("Math")
+		tolerance := 1.0e-5f
+		this add("Modulo", func {
+			// Note: Negative values are in parentheses because of bug in rock
+			expect(22 modulo(5), is equal to(2))
+			expect((-7) modulo(3), is equal to(2))
+			expect((-7) modulo(1), is equal to(0))
+			expect(0 modulo(3), is equal to(0))
+			expect(3 modulo(4), is equal to(3))
```
Comment (simonmika) [inline test/sdk/MathTest.ooc]: Thanks for the additional tests.

**Comment by marcusnaslund** [comment]
Cleaned up some parentheses. The rest of the comments will be taken care of in a separate cleanup PR of the entire `math.ooc` file.

**Comment by simonmika** [comment]
Thanks, peer review done.
